### PR TITLE
vm: setup mantle hardfork(protocol upgrade) and add new PUSH0 opcode

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes

- setup mantle hardfork (not activated yet. it would activate in later PR.) and add PUSH0 opcode.
  - PUSH0 is part of shanghai ethereum hardfork(https://eips.ethereum.org/EIPS/eip-3855)
- tested with a contract compiled by solidity v0.8.20. It contains PUSH0 opcode.
  - https://blog.soliditylang.org/2023/05/10/solidity-0.8.20-release-announcement/ 
  - result
    - when mantle is not activated
      - the test contract deployment is failed. (PUSH0 opcode is not processible)
    - when mantle is activated
      - the test contract deployment is succeed.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by blukat29 — DISMISSED** [review]

File: blockchain/vm/eips.go [hunk 1]
```diff
@@ -188,3 +191,20 @@ func enable4399(jt *JumpTable) {
...
 }
+
+// enable3855 applies EIP-3855 (PUSH0 opcode)
+func enable3855(jt *JumpTable) {
+	jt[PUSH0] = &operation{
+		execute:         opPush0,
+		constantGas:     GasQuickStep,
+		minStack:        minStack(0, 1),
+		maxStack:        maxStack(0, 1),
+		computationCost: params.PushComputationCost,
```
Comment (hyunsooda) [inline blockchain/vm/eips.go:202]: I think the introduced function's running time is much faster and cheaper. Gas cost can be saved to utilize this instruction, but not affected to the computation cost in this change. They should have correleation each other. For example, if the gas is consumed bigger, then the computation cost also should be increased, and vice versa.

**Review by hyunsooda — COMMENTED** [review]
This follow-up PR is commendable. I have verified that incorporating the `PUSH0` instruction significantly reduces the contract size, making it an excellent feature.

File: blockchain/vm/eips.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/vm/eips.go:202]: @hyunsooda Good point. It is said the performance is improved a lot. You know, we usually use [this benchmark test](https://github.com/klaytn/klaytn/blob/46b075f5a83578cef74cf3d6015c64f213a23175/blockchain/vm/instructions_test.go#L1011-L1013) to decide computation cost. However, the result seems not improved. The result varies a lot.
```
➜  vm git:(PUSH0) ✗ go test -bench=BenchmarkOpPush0 -benchtime 200x
goos: darwin
goarch: arm64
pkg: github.com/klaytn/klaytn/blockchain/vm
BenchmarkOpPush0-10          200                30.83 ns/op

➜  vm git:(PUSH0) ✗ go test -bench=BenchmarkOpPush1 -benchtime 200x
goos: darwin
goarch: arm64
pkg: github.com/klaytn/klaytn/blockchain/vm
BenchmarkOpPush1-10                  200                41.25 ns/op 
```

Moreover, i just use PUSH Computation cost because there's an issue that our computation cost measuring way is not exact.

What do you think about it?

Comment (hyunsooda) [reply blockchain/vm/eips.go:202]: When comparing the actual implementation code, there is a notable difference between the opcode `PUSH0` and the `makePush` function. The `PUSH0` opcode consists of a single line that initializes the big integer type, while the `makePush` function comprises multiple instructions.

For instance, let's assume there are 100 occurrences of the `PUSH0` opcode in a contract. In terms of computational cost (CC), these `PUSH0` instructions are considered to have the same cost as other `PUSHX` instructions. However, it's important to note that the CC value is not saved, which can potentially result in transaction failures although the actual CC of `PUSH0` is cheaper. This discrepancy in CC calculation can lead to unintended issues.

Furthermore, it's essential to consider the lifetime of the contract. The contract remains active unless its code is destroyed using the `selfdestruct`. Therefore, the aforementioned computational cost must be paid continuously throughout the contract's existence. This is my overall concerns.

In summary:
- Decreased gas cost should be accompanid the decreased CC as well.
- Consider the contract lifetime and running the same multiple instructions.

Conclusion:
I believe the appropriate solution for this particular PR would be to reduce the CC associated with the `PUSH0` opcode. While addressing the removal of CC and improving its measurement is a separate topic that can be discussed at a later time.

Please feel free to add further comment.

File: params/config.go [hunk 2]
```diff
@@ -419,6 +431,12 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, head *big.Int) *Confi
 	if isForkIncompatible(c.KoreCompatibleBlock, newcfg.KoreCompatibleBlock, head) {
 		return newCompatError("Kore Block", c.KoreCompatibleBlock, newcfg.KoreCompatibleBlock)
 	}
+	if isForkIncompatible(c.Kip103CompatibleBlock, newcfg.Kip103CompatibleBlock, head) {
```
Comment (ian0371) [inline params/config.go:434]: I thought this was left intentionally blank because kip103 is an optional fork. @aidan-kwon is this ok?

File: params/config.go [hunk 3]
```diff
@@ -381,6 +391,8 @@ func (c *ChainConfig) CheckConfigForkOrder() error {
 		{name: "ethTxTypeBlock", block: c.EthTxTypeCompatibleBlock},
 		{name: "magmaBlock", block: c.MagmaCompatibleBlock},
 		{name: "koreBlock", block: c.KoreCompatibleBlock},
+		{name: "kip103Block", block: c.Kip103CompatibleBlock},
```
[the remark this answers is not in the document]
Comment (yoomee1313) [reply params/config.go]: @ian0371 I didn't know it was intended.. Thanks for letting me know about it.

[the remark this answers is not in the document]
Comment (ian0371) [reply params/config.go]: It's what I thought initially so I need double check...

[the remark this answers is not in the document]
Comment (yoomee1313) [reply params/config.go]: If it is an optional fork, it looks natural to me not to check this fork ordering :)

File: blockchain/vm/eips.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/vm/eips.go]: @hyunsooda Okay, I'll reduce the computation cost to 2/3 level of the other PUSH computation cost value. (120 -> 80) https://github.com/klaytn/klaytn/pull/1861/commits/1cc1ab0703502924ba4b318f603b8836b9e6f299.

File: params/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply params/config.go]: @yoomee1313 We've intentionally missed kip103Block in the fork ordering check because kip103 is designed as an optional hardfork and there are no dependence with other forks.

[the remark this answers is not in the document]
Comment (aidan-kwon) [reply params/config.go]: We may leave a comment in our source code for the future developers who doesn't know about the context

**Review by blukat29 — DISMISSED** [review]

**Review by hyunsooda — DISMISSED** [review]

**Review by ian0371 — APPROVED** [review]

**Review by blukat29 — APPROVED** [review]
